### PR TITLE
Add onboarding flow state persistence

### LINKED ISSUE #5477: Restore Onboarding Flow State
### Description

We may want to allow a user to leave the new onboarding flow process and return to it later. This could be done for example by saving in the account meta, or an option cache what the user has done so far, so that they can pick up the onboarding again at a later date. This only covers the new onboarding flow form, including its context and the `Stepper` one.

### Acceptance criteria

* User can leave the onboarding and come back.
* It will remember where they were in the onboarding, and prefill what they have already filled out.

### PR DESCRIPTION
**Pull request by ismaeldcom**
Fixes #5477 

#### Changes proposed in this Pull Request
- Add `get`, `set`, and `clear` onboarding flow state option methods in the Onboarding Service.
- Add editable `/onboarding/flow-state` REST route to update the option value from the JS client.
- Inject option value under `wcpaySettings.onboardingFlowState`.
- Update `<Stepper/>` to support `initialStep` prop.
- Persist state using `onStepChange` event, to ensure it's only persisted once per each step with valid data.
- Recover state using both `initialStep` and `intialData` only if `wcpaySettings.onboardingFlowState` exists.
- Clear the onboarding flow state on Stripe onboarding initialization, to avoid keeping an old state.
- Add a dismissable info banner, when there is a persisted state, to inform the merchant about it.

- With PO enabled go to the onboarding flow.
- Pick live onboarding option.
- Fill in the personal details, and click on **Continue**.
- Refresh the page.
- You should land on the business step, just where you left it, and the new notice banner should be visible.
- Go back using the navbar button.
- Your previous details should be there.
- Finish the onboarding process.
- Start again, and confirm that the persisted state is gone.

#### Screenshot – New info banner
Screenshot 2023-08-15 at 09 26 13

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6934` or branch name `add/5477-onboarding-progress-persistence` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/5477-onboarding-progress-persistence) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 80dead31e1d238c43dfc357feaef2c8a3d5eae4c
- Build time: 2023-08-23 07:58:04 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by ismaeldcom** [comment]
>Add a dismissable info banner, when there is a persisted state, to inform the merchant about it.

 @elizaan36 could you please review this part? There is an attached screenshot on the PR description. We had a quick conversation about adding this on Slack (p1691056691218109-slack-C03KTTK2YMA) But feel free to suggest an alternative approach if you like 🙂

File: client/onboarding/strings.tsx [hunk 1]
```diff
@@ -187,6 +187,10 @@ export default {
 		from_3_to_6months: __( '3 – 6 months', 'woocommerce-payments' ),
 		more_than_6months: __( '6+ months', 'woocommerce-payments' ),
 	},
+	restoredState: __(
+		'We have restored your previous session. You can pick up where you left off or go back to a previous step to make changes. ',
```
Comment (dmallory42) [inline client/onboarding/strings.tsx]: Nitpick: what do you think about adding a comma to this sentence:

```You can pick up where you left off, or go back to a previous step to make changes.```

Even though it's grammatically correct either way, I think the comma helps present the options as two distinct choices (go back or continue). 🙂 Happy to go with whatever you think is best!

**Review by dmallory42 — APPROVED** [review]
LGTM! Tested and works perfectly, and I really like the approach taken here. One minor comment for you to look at but I'm pre-approving so we can :shipit: once the design team have had a look!

File: client/onboarding/strings.tsx (hunk 1 above)
Comment (ismaeldcom) [reply client/onboarding/strings.tsx]: Actually I added the comma on an earlier version, not sure why I removed it later 🤔 Thanks for catching it! _Addressed in 738f73b_

**Comment by elizaan36** [comment]
@ismaeldcom Sorry I missed this one. We should use the `Banner notice` component here rather than `Inline notice` since it's at the page top level. I shared the Design Library Figma in Slack.

image

One other thing, I think it would look visually aligned if the notice width extends to the width of the body text container below. In your screenshot the notice looks slightly narrower.

The copy looks good to me.

**Comment by ismaeldcom** [comment]
Thanks for the review @elizaan36. Probably the confusion comes from our existing `BannerNotice` component which actually matches the `Inline notice` one in Figma 🙂 As this change will require an update of multiples places where the current one is used, I've just created a new issue to handle both the creation and rename #7048

I'll merge this one, with the fixed margin, so it takes the full available width. And jump to the next issue to ensure it lands on the next release with the proper design.

**Pull request merged**